### PR TITLE
Add the Mesquite mesh adaption library

### PR DESCRIPTION
**Pull request by jppelteret**
This initial commit adds the Mesquite mesh adaption library to our build system and demonstrates its use. I am in the process of writing an interface to this library, so there's obviously more to come.

Note: The small VTK test meshes added to this PR are courtesy of [the Mesquite project](https://github.com/trilinos/mesquite).

### CODE REVIEW

File: tests/mesquite/basic_01.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * See sec 3.1.3 Improving the Mesh with a Wrapper Class
+ */
+
+
+#include "../tests.h"
+
+#include <deal.II/mesquite/mesquite.h>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
```
Comment (davydden) [inline tests/mesquite/basic_01.cc]: you probably need to remove this.

File: tests/mesquite/basic_02.cc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * See sec 3.1.4 Improving the Mesh with the Low Level API
+ */
+
+
+#include "../tests.h"
+
+#include <deal.II/mesquite/mesquite.h>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
```
Comment (davydden) [inline tests/mesquite/basic_02.cc]: same

File: tests/mesquite/basic_03.cc [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+ *
+ * See sec 4.2 Accessing Mesh In Arrays
+ */
+
+
+#include "../tests.h"
+
+#include <deal.II/mesquite/mesquite.h>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
```
Comment (davydden) [inline tests/mesquite/basic_03.cc]: same

File: tests/mesquite/basic_01.cc [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ * Mesquite Mesh Quality Improvement Toolkit User's Guide
+ * Sandia National Laboratories, Sandia National Laboratories, 2013
+ *
+ * See sec 3.1.3 Improving the Mesh with a Wrapper Class
+ */
+
+
+#include "../tests.h"
+
+#include <deal.II/mesquite/mesquite.h>
```
Comment (davydden) [inline tests/mesquite/basic_01.cc]: i think you forgot to add this one, I can't find it

Comment (jppelteret) [reply tests/mesquite/basic_01.cc]: Indeed I did. I've now added this file.

**Comment by tamiko** [comment]
@jppelteret Can you quickly shed some light on this one: I see that trilinos has a mesquite project
```
% qlist trilinos | grep mesquite
/usr/include/trilinos/mesquite_config.h
/usr/include/trilinos/mesquite_version.h
/usr/lib64/trilinos/libmesquite.so.12
/usr/lib64/trilinos/libmesquite.so
/usr/lib64/trilinos/libmesquite.so.12.10.1
```
But there is also an external mesquite project (from sandia):
```
% eix mesquite
* sci-libs/mesquite [1]
     Available versions:  ~2.3.0
     Homepage:            http://www.cs.sandia.gov/optimization/knupp/Mesquite.html
     Description:         Mesh Quality Improvement Toolkit
```

Do they bundle a copy? Or is this an unrelated project?
I guess you want to pick up the latter one (judging from the cmake files.)

**Comment by tamiko** [comment]
If so, would you mind adding a small blurbb to the cmake configuration to extract the version number? :stuck_out_tongue_closed_eyes:

**Comment by davydden** [comment]
> I see that trilinos has a mesquite project

i guess that's specifics of your Linux, AFAIK `trilinos` does not bundle `mesquite` anymore, see https://github.com/trilinos/Trilinos/tree/master/packages 

I also don't have it in my installation 
```
$ ls ~/spack/opt/spack/darwin-highsierra-x86_64/clang-9.0.0-apple/trilinos-12.12.1-itckqyn2f4wc2lujxeffvoq7w4aamecl/lib/ | grep mesquite
```

**Comment by jppelteret** [comment]
@tamiko What @davydden said. At some point in time they stopped bundling it (I was surprised to find that out)... I'll add a mechanism to pick up the version number.

**Comment by jppelteret** [comment]
@tamiko Done.

**Comment by tamiko** [comment]
That's a bit surprsigin.
The current version of Trilinos is 12.12.1 - according to their website [1] they still package mesquite.

[1] https://trilinos.org/download/

**Comment by davydden** [comment]
@tamiko i checked my build logs and I don't see `mesquite` there:
```
Final set of enabled packages:  Gtest Kokkos Teuchos KokkosKernels Sacado Epetra Zoltan Triutils Tpetra TrilinosSS EpetraExt Xpetra AztecOO Galeri Amesos Zoltan2 Ifpack ML Belos Amesos2 SEACAS Anasazi Ifpack2 MueLu ROL 25

Final set of enabled SE packages:  Gtest KokkosCore KokkosContainers KokkosAlgorithms Kokkos TeuchosCore TeuchosParameterList TeuchosComm TeuchosNumerics TeuchosRemainder TeuchosKokkosCompat TeuchosKokkosComm Teuchos KokkosKernels Sacado Epetra Zoltan Triutils TpetraClassic TpetraTSQR TpetraCore Tpetra TrilinosSS EpetraExt Xpetra AztecOO Galeri Amesos Zoltan2 Ifpack ML Belos Amesos2 SEACASExodus SEACASExodus_for SEACASExoIIv2for32 SEACASNemesis SEACASIoss SEACASChaco SEACASAprepro_lib SEACASSupes SEACASSuplib SEACASSuplibC SEACASSuplibCpp SEACASAlgebra SEACASAprepro SEACASConjoin SEACASEjoin SEACASEpu SEACASExo2mat SEACASExodiff SEACASExomatlab SEACASExotxt SEACASExo_format SEACASEx1ex2v2 SEACASGjoin SEACASGen3D SEACASGenshell SEACASGrepos SEACASGrope SEACASMapvarlib SEACASMapvar SEACASMapvar-kd SEACASMat2exo SEACASNemslice SEACASNemspread SEACASNumbers SEACASTxtexo SEACASEx2ex1v2 SEACAS Anasazi Ifpack2 MueLu ROL 74

Final set of non-enabled packages:  ThreadPool RTOp MiniTensor Shards GlobiPack Domi Thyra OptiPack Isorropia Pliris Claps Pamgen ShyLU Trios Komplex Stratimikos FEI Teko TriKota Intrepid Intrepid2 STK Phalanx NOX Moertel Rythmos Tempus Stokhos Piro Panzer PyTrilinos NewPackage MeshingGenie TrilinosCouplings Pike 35

Final set of non-enabled SE packages:  ThreadPool KokkosExample TeuchosParser RTOp MiniTensor Shards GlobiPack Domi ThyraCore ThyraEpetraAdapters ThyraEpetraExtAdapters ThyraTpetraAdapters Thyra OptiPack Isorropia Pliris Claps Pamgen ShyLUHTS ShyLUTacho ShyLUBDDC ShyLUBasker ShyLUFastILU ShyLUCore ShyLU SEACASSVDI SEACASPLT SEACASBlot SEACASFastq Trioscommsplitter Triossupport Triosnnti Triosnssi Triosprograms Triosexamples Triostests Triosnetcdf-service Trios Komplex Stratimikos FEI Teko TriKota Intrepid Intrepid2 STKClassic STKUtil STKSimd STKTopology STKMesh STKIO STKUnit_test_utils STKSearch STKSearchUtil STKTransfer STKUnit_tests STKDoc_tests STKExp STKExprEval STK Phalanx NOX Moertel Rythmos Tempus Stokhos Piro PanzerCore PanzerDofMgr PanzerDiscFE PanzerAdaptersSTK Panzer PyTrilinos NewPackage MeshingGenie TrilinosCouplings PikeBlackBox PikeImplicit Pike 79

Final set of enabled TPLs:  Pthread MPI BLAS LAPACK Boost METIS ParMETIS HDF5 Netcdf SuperLUDist UMFPACK HYPRE SCALAPACK MUMPS Matio DLlib 16

Final set of non-enabled TPLs:  MKL yaml-cpp Peano CUDA CUSPARSE Thrust Cusp TBB HWLOC QTHREAD BinUtils ARPREC QD Scotch OVIS gpcd PuLP TopoManager LibTopoMap PaToH CppUnit ADOLC ADIC TVMET MF ExodusII Nemesis XDMF Zlib CGNS Pnetcdf y12m SuperLUMT SuperLU Cholmod MA28 AMD CSparse PETSC BLACS PARDISO_MKL PARDISO Oski TAUCS ForUQTK Dakota HIPS MATLAB CASK SPARSKIT QT gtest BoostLib BoostAlbLib OpenNURBS Portals CrayPortals Gemini InfiniBand BGPDCMF BGQPAMI Pablo HPCToolkit Clp GLPK qpOASES PAPI MATLABLib Eigen X11 Lemon GLM quadmath CAMAL RTlib AmgX CGAL CGALCore VTune TASMANIAN ArrayFireCPU SimMesh SimModel SimParasolid SimAcis SimField 86
```

**Comment by jppelteret** [comment]
@tamiko I'm in the same boat as @davydden that my Trilinos does not have this packaged. I suppose there may then be a conflict if Trilinos is built with Mesquite and then we specify that an external Mesquite be used as well. 

Would you (since you've got a Trilinos with packaged Mesquite) be willing to amend the CMake module to (i) detect if there's a working Trilinos with a bundled Mesquite, and if so (ii) use that in preference to the external package?

**Comment by tamiko** [comment]
@jppelteret Can do that. Let me think about it this afternoon :-)

**Comment by jppelteret** [comment]
Thank you CMakemeister :-)

**Comment by Rombur** [comment]
Mesquite is an external package of Trilinos and I think that it is added to the tarball from the trilinos werbsite when the release is done. However, spack is using the tarball made github and so it is not part of it.

**Comment by davydden** [comment]
@Rombur thanks for clarification, then we indeed need to support both cases.

**Comment by tamiko** [comment]
@jppelteret Whether that's a compliment has yet to be decised :stuck_out_tongue:

**Comment by jppelteret** [comment]
@tamiko Whether or not it was meant as a compliment is yet to be decided 😁 

If I understand correctly, `trilinos` is guarenteed to be configured before `mesquite`. So would it suffice if we detect `mesquite` as an optional module in `trilinos`, and based on whether the `TRILINOS_WITH_MESQUITE` flag is set choose to (i) continue with another external `mesquite` configuration or (ii) just set `DEAL_II_WITH_MESQUITE` to true?

**Comment by tamiko** [comment]
> If I understand correctly, `trilinos` is guarenteed to be configured
> before `mesquite`. So would it suffice if we detect `mesquite` as an
> optional module in `trilinos`, and based on whether the
> `TRILINOS_WITH_MESQUITE` flag is set choose to (i) continue with
> another external `mesquite` configuration or (ii) just set
> `DEAL_II_WITH_MESQUITE` to true?

Yes, something on that line.
Only complication is that the current feature configuration doesn't
support that and I want to think about a sane way of integrating
something like that.

**Comment by jppelteret** [comment]
@davydden I think the issue is that this could lead to broken builds (if `mesquite` is detected by both this module and `trilinos`), and so it should not be merged.

**Comment by davydden** [comment]
> I think the issue is that this could lead to broken builds (if mesquite is detected by both this module and trilinos)

I don't think it will happen now. You are not checking `mesquite` in Trilinos, so even if it is compiled with `trilinos`, you won't get it as a part of our `trilinos` configuration. So it should not harm, it just does not support both scenarios yet.

**Comment by jppelteret** [comment]
Oh yeah, that's true. It is [not in our module list](https://github.com/dealii/dealii/blob/master/cmake/configure/configure_2_trilinos.cmake#L45) and each trilinos package is built as a seperate librar,y so it shouldn't appear in the link line. Lets wait for a final confirmation from @tamiko before proceeding though.

**Comment by jppelteret** [comment]
Implementing some tests has lead me to discover that I've introduced a namespace clash here: `::Mesquite` (and `::Mesquite2` are already used by `mesquite`. Should I change the namespace I've used here, and if so then are there any suggestions as to what I could change this to? I was thinking that maybe `MeshAdaption` might be appropriate.

**Comment by davydden** [comment]
> I was thinking that maybe MeshAdaption might be appropriate.

We already have `GridTools` (with `laplace_transform`, so maybe use it but define things in a separate header?

**Comment by davydden** [comment]
> I was thinking that maybe MeshAdaption might be appropriate.

We already have `GridTools` (with `laplace_transform`, so maybe use it but define things in a separate header?

**Comment by davydden** [comment]
or even put in the same header `deal.II/grid/grid_tools.h` and guard with `#ifdef ... #endif`.

**Comment by jppelteret** [comment]
I agree that the `GridTools` namespace is probably a better place for this, so I'll make the necessary changes shortly. I think, though, that I'd put this in another file (maybe `grid_tools_mesquite.h` since `grid_tools.h` is pretty sizable as it is. Then at least this is self-contained and hopefully more maintainable.

**Comment by davydden** [comment]
> I'd put this in another file (maybe grid_tools_mesquite.h since grid_tools.h is pretty sizable as it is.

makes perfect sense to me.

**Comment by jppelteret** [comment]
Ok, so I'll mark this as a WIP while I make the changes.

**Comment by bangerth** [comment]
On 03/01/2018 05:29 AM, Jean-Paul Pelteret wrote:
> I agree that the |GridTools| namespace is probably a better place for this, so 
> I'll make the necessary changes shortly. I think, though, that I'd put this in 
> another file (maybe |grid_tools_mesquite.h| since |grid_tools.h| is pretty 
> sizable as it is. Then at least this is self-contained and hopefully more 
> maintainable.

How much material are you going to add to the header file? If it's just a 
small number of functions, I actually like the idea of putting them into the 
same header because then you don't have to remember which header file needs to 
be included in user projects.

That does not preclude you from putting the implementation into a separate .cc 
file, of course.

**Comment by bangerth** [comment]
On 03/01/2018 03:32 AM, Jean-Paul Pelteret wrote:
> Implementing some tests has lead me to discover that I've introduced a 
> namespace clash here: |::Mesquite| (and |::Mesquite2| are already used by 
> |mesquite|.

There should be no clash if you put your namespace into namespace dealii. It 
just forces you to write ::Mesquite::function_name() or 
::dealii::Mesquite::function_name() whenever you reference a function.

**Comment by jppelteret** [comment]
> How much material are you going to add to the header file?

The current header size, with documentation but without inline definitions, weighs in at ~750 lines. So not too big, I suppose. But `grid_tools.h` is already over 3000 lines long.

> There should be no clash if you put your namespace into namespace dealii.

Well, here's an test using my first version of my current implementation
```c++
  // Perform mesh smoothing
  dealii::Mesquite::MeshInterface<dim> mesquite_interface (tria, true /*fix_all_boundary_vertices*/);
  mesquite_interface.execute(dealii::Mesquite::WrapperTypes::shape_improvement);
  // Move the triangulation's vertices
  mesquite_interface.move_vertices(tria);
```
and this is the new one with the classes and functions now put into `GridTools`
```c++
  // Perform mesh smoothing
  GridTools::MesquiteMeshInterface<dim> mesquite_interface (tria, true /*fix_all_boundary_vertices*/);
  mesquite_interface.execute(GridTools::MesquiteWrapperTypes::shape_improvement);
  // Move the triangulation's vertices
  mesquite_interface.move_vertices(tria);
```

There's not much in it in terms of line length, but I think that the second looks better and is more convenient (natural?) not to have to explicitly state the `deal.II` namespace. 

@dealii/dealii What do you think? Vote for the first with a 😄  and the second with a 🎉 . I don't know why, but give me a ❤️ too while your mouse is near that icon 😝

**Comment by davydden** [comment]
@jppelteret that's a very clever way of making a poll! 👍

**Comment by kronbichler** [comment]
> The current header size, with documentation but without inline definitions, weighs in at ~750 lines. So not too big, I suppose. But grid_tools.h is already over 3000 lines long.

How about you put it in a separate header file that gets then included in `grid_tools.h`? I know it's maybe not the best style to mix a big header file with its own definition with unconditional includes of additional header files just for the purpose of getting a unified interface. But some of the big headers like `vector_tools.h` would definitely profit from some additional structure, i.e., do like boost is doing with one collective header that only pulls in other headers by sub-groups.

**Comment by jppelteret** [comment]
@davydden Not my idea - someone else (I can't remember who) did something similar a few weeks ago.

**Comment by masterleinad** [comment]
> @davydden Not my idea - someone else (I can't remember who) did something similar a few weeks ago.

I guess @drwells came up with this idea in #5889.

**Comment by jppelteret** [comment]
> How about you put it in a separate header file that gets then included in grid_tools.h?

Good idea @kronbichler. I know for a fact that @davydden agrees with this approach, because he suggested it to me about 2 minutes ago :-) I'd be keen for us to do this.

**Comment by davydden** [comment]
>  https://github.com/apex/gh-polls seems to provide voting for GitHub projects, but I know nothing more about it other than that it exists.

looks like one can simply install the app into the GitHub repo, from https://gh-polls.com

> Polls can be created via the polls command-line tool, via the web application, or install the GitHub App in your repository.

https://github.com/apps/polls

**Comment by bangerth** [comment]
If most of the 750 lines is documentation, then the extra lines don't bother me too much.

I'm not a fan of putting things into a separate header file that is then unconditionally `#include`d. What would be the purpose of this? From a user perspective, it's the same as having the additional lines directly in the original header file. From a developer perspective, you still have to remember which file things are in. The only advantage is that you wouldn't have to remember where exactly in `grid_tools.h` the Mesquite functions are -- closer to the top, closer to the bottom, etc. But that's only an issue if you use vi or emacs, and then I have no patience for you anyway. Just go use Eclipse or Netbeans or QtCreator, or anything else that cares about *content* and not *layout*. I never scroll up and down to a particular line in files any more -- I just hit `Ctrl-O`, which opens a dialog with all functions, variables, and classes in a file, and then start typing the name of the declaration I want to jump to. I don't need to know (and typically don't actually know) any more where exactly in a file a particular function is defined -- Eclipse will take me there when I need to go there.

The only reason to do this would be if it allowed us to reduce compile times, for example if some of the .cc files only included *parts* of `vector_tools.h` and `vector_tools.templates.h` that have been externalized into separate files. But even that is of questionable use: doxygen documents at the top of each class and namespace page which header file stuff resides in, and I bet it would be confusing to users to have to see a half dozen separate header files for the functions in `VectorTools`.

**Comment by kronbichler** [comment]
My point was not to avoid big files for layout reasons, but really for `content` reasons. The things in `VectorTools` do many different things and additional files could be one way of grouping the different functionality. My impression is that many students are overwhelmed by the big size of the files, of the documentation pages and just miss existing functionality. But it's true that there also is an opposite effect of not knowing where to find things when there is then places to look for. So let's not discuss this too much, I guess we have slightly different opinions locally but pretty much the same view globally.

As to the editor issue I cannot help but express a somewhat opposite opinion. Emacs and vi mostly do not really care whether your file is 200 lines or 200,000 lines. It's with these editors I prefer big files. They are always quick, search+replace is easier within a file, etc. On the other hand, at least QtCreator and Eclipse struggle with files that have more than a few thousands of lines and you get this nasty lag while you're trying to do something. Scalability mode in Eclipse is there for a reason, even if you can change the threshold. All file outline of functions gets pretty bad on big files with unrelated functionality. I don't know about what is different for Eclipse to have content knowledge across the whole project and to what is displayed within a file, but it must be immense. And you get this especially when you scroll back and forth, which is common at least in my workflow. Not in the initial stage of locating something, but when actually implementing or trying to figure out where that nasty bug could be. And just to clarify, I am really in that same boat: I do my serious programming with QtCreator and many of my students use Eclipse.

Still I definitely wouldn't diminish the value of vi/emacs also against these alternatives: On machines with remote access I would always prefer `vi` or `emacs -nw` over a context-oriented editor. Well, I did run the front end locally on the files of an ssh-fs-mounted path more than once to get more - but I have too many "short" projects on different machines where firing up the whole toolchain with remote compilation and stuff like that is simply not better than just rsyncing to the local machine. Or using vi/emacs.

**Comment by kronbichler** [comment]
Anyway, now I've diverged a bit :smile: but I agree with the suggestion by @jppelteret and have no objection to have it in the same file.

**Comment by jppelteret** [comment]
Thank you all for the lengthy discussion on where to put this. I've come up with an intermediate solution while I'm still implementing this - I've simply removed the main library header and have temporarily moved the exception that I placed in it into the test utility file. I reckon though that this PR should only be merged after the next release - that would give time for @tamiko to deal with the build system, and I need more time to work through the finer details of the implementation.

In the mean time, here's a demonstration of progress...

Consider a problem with a reference configuration that looks like this
reference

but a displacement map that leads to this distorted mesh
distorted

We can then untangle and adapt deformation map like this
smoothed

and provide the opportunity to do any number of things with the corrected / optimised displacement map, such as move the (reference) triangulation vertices, pass it back to the user for other use, etc.

**Comment by bangerth** [comment]
@jppelteret -- Current status?

**Comment by bangerth** [comment]
@jppelteret -- was this PR ready all along and we just dropped the ball in reviewing/merging it? Or was there anything that still needed to be done?

**Comment by jppelteret** [comment]
> was this PR ready all along and we just dropped the ball in reviewing/merging it? Or was there anything that still needed to be done?

@bangerth I've just spotted your ping. I guess that this could be merged. Skimming over our previous discussions, it would seem like the concern of a clash between a Trilinos-bundled version of Mesquite and a stand-alone library is (maybe) resolved? Should I rebase the branch on master to ensure that the indentation of the tests is fine?

**Comment by bangerth** [comment]
Yes, do that if you don't mind!

**Comment by bangerth** [comment]
One of the testers fails, but I think for unrelated reasons...

File: cmake/modules/FindMESQUITE.cmake [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ENDIF()
+
+
+DEAL_II_PACKAGE_HANDLE(MESQUITE
+  LIBRARIES 
+    REQUIRED MESQUITE_LIBRARY
+    OPTIONAL MESQUITE_UTILITY_LIBRARY
+  INCLUDE_DIRS REQUIRED MESQUITE_INCLUDE_DIR
+  USER_INCLUDE_DIRS REQUIRED MESQUITE_INCLUDE_DIR
+  CLEAR MESQUITE_INCLUDE_DIR MESQUITE_LIBRARY MESQUITE_UTILITY_LIBRARY
```
Comment (bangerth) [inline cmake/modules/FindMESQUITE.cmake]: This file doesn't produce any output explicitly, but I don't know whether `DEAL_II_PACKAGE_HANDLE` outputs anything. Either way, can you make sure that you print that Mesquite was found and its version identifier into the output?

**Comment by bangerth** [comment]
@jppelteret -- What is the status here?

**Comment by drwells** [comment]
This is actually not that far from done - the function `smooth_mesquite_in_place_arraymesh()` in the test header does most of what we want.

Does anyone want to volunteer to wrap this up?

**Comment by jppelteret** [comment]
Sorry, I'm still not finding any time to contribute anything.

I think that since we now have the capabilities to select optional packages from Trilinos, we should dump all of the CMake stuff and just make sure that Mesquite comes from Trilinos. Its a bit of a pity since Mesquite is relatively lightweight, but I think that its the right (or easiest) thing to do.

> the function smooth_mesquite_in_place_arraymesh() in the test header does most of what we want.

That was the starting point that I used for #8196, which naturally now gives a more flexible interface. I believe that one might be able to support fixing vertices on manifolds using the `[add/set]_mesh_fixed_vertices()` function. Mesquite doesn't support real constraints, which is a bit of a pity.

**Comment by jppelteret** [comment]
@drwells I'm busy reinstalling my toolchain and dependencies now. I'm going to see if I can get Mesquite bundled with Trilinos to be picked up, and will try to finish this myself.

**Comment by drwells** [comment]
Glad to hear it! :)

**Comment by jppelteret** [comment]
@drwells @tamiko I tidied this up, making Mequite a Trilinos optional dependency. When you have the time, would you mind please giving it a look?

**Comment by drwells** [comment]
I'll find some time over the next week to take a look.

**Comment by drwells** [comment]
I recompiled Trilinos 12.18 and it looks like it doesn't include mesquite. The most recent commits in the repo are from 2018, 2018, 2018, and 2014 - I think this project is dead. Are we sure we want to proceed with this? Is there a replacement for mesh quality improvement in Trilinos?

**Comment by drwells** [comment]
@jppelteret How did you compile this? I don't know enough about tribits to compile the library separately.

**Comment by jppelteret** [comment]
@drwells I've got a [PR out](https://github.com/spack/spack/pull/14723) to add this to spack. Basically, you just need to extract mesquite into a folder `packages/mesquite` (i.e. alongside all of the other bundled projects) and continue as per usual. I don't really have an opinion as to whether we should be looking for this as a part of Trilinos or as an external project. From what @tamiko said earlier, it appears to still be bundled into some distro's package. Looking [here](https://packages.ubuntu.com/source/xenial/trilinos), it seems like its not a part of the Trilinos that comes with Ubuntu 16.04 LTS? I don't mind putting back my CMake module to find Mesquite as a separate project, but we would then need to find a way to address the possible ambiguity that @tamiko mentioned might arise.

But I I think that you observation is correct -- that its a "mature" project, and is possibly no longer maintained. I didn't find any alternatives to Mesquite when looking for a tool to do this job. If we collectively think that its a bad idea to incorporate it into the library, then I'd understand. But I _think_ that it still finds use in Cubit/Trellis, at least.

**Comment by jppelteret** [comment]
Oh, and I've also got [this PR](https://github.com/trilinos/mesquite/pull/4) against Mesquite that's been sitting around unnoticed for 2 years. I used that patch to build it as a stand-alone library on my Mac.

**Comment by drwells** [comment]
I agree that having a mesh quality improvement tool would be useful but I don't think its a good idea to add things that require special steps to re-add to trilinos. What do others think?

**Comment by Rombur** [comment]
> I agree that having a mesh quality improvement tool would be useful but I don't think its a good idea to add things that require special steps to re-add to trilinos. 

Why not? I don't see what's the problem with that. If it's the way it is supposed to be built, then that's the way it is.

**Comment by drwells** [comment]
For example, we will have to change all the CI build scripts to install trilinos in an unusual way - we won't be able to pull in all the Trilinos packages with a default installation anymore.

**Comment by Rombur** [comment]
That's not really hard to do though. Just go in Trilinos/package and pull Mesquite. I am using another package that is installed like Mesquite, it looks strange but it's actually pretty easy.

**Comment by luca-heltai** [comment]
If possible, I'd vote to have both options: see if it's in trilinos, then enable it and forget about it, or see if it is external.
I'm thinking about situations where Trilinos cannot be installed easily, while mesquite can (windows, for example). I would not want to add an unnecessary dependency on something that is not that trivial/light/etc. to install.

**Comment by bangerth** [comment]
Unrelated to this patch, but if we're worried whether or not Mesquite is maintained or not, we could ask the authors of that package!

**Comment by jppelteret** [comment]
Its annoying that both Mesquite and Verdict (a part of the VTK library) are bundled together with such heavy packages. I managed to get them out and to be self-contained, but it doesn't seem like its of any interest upstream.

**Comment by Rombur** [comment]
> Unrelated to this patch, but if we're worried whether or not Mesquite is maintained or not, we could ask the authors of that package!

I've asked and this project is dead. The main developer retired and they didn't have any customers at Sandia.

**Comment by bangerth** [comment]
Aw, bummer :-(

**Comment by jppelteret** [comment]
I guess we're now back at [this point](https://github.com/dealii/dealii/pull/5971#issuecomment-369482313) and the comments/questions preceding it, right?

**Comment by jppelteret** [comment]
Or do we need to discuss whether or not its a good idea to integrate features from a "dead" project into the library?

**Comment by bangerth** [comment]
Could we ingest Mesquite into `bundled/`? I have no problem with linking against dead libraries as long as we can be reasonably certain that they can be found online for the foreseeable future. But if we have it in bundled/, then we can make sure it compiles cleanly with all compilers we support etc.

**Comment by Rombur** [comment]
> Could we ingest Mesquite into bundled/? 

Sounds good to me and the good thing is that we will never need to update it :smile:

**Comment by jppelteret** [comment]
> Could we ingest Mesquite into `bundled/`?

Yes, I'll look at integrating it into `bundled` in the coming days. Its quite small once one drops the testsuite & its associated mesh files. We might want to consider doing the same for Verdict, when we get there.

> we will never need to update it

After I patch it to fix a broken feature :-D

**Comment by drwells** [comment]
The last time I checked it sounded like people were okay with bundling this. @jppelteret are you still willing to do that work?

**Comment by jppelteret** [comment]
Yip. I think that I'm getting pretty close. I need to rewrite `CMakeLists.txt` for Mesquite, as is done for the other projects.

**Comment by jppelteret** [comment]
@drwells Closer, but still not quite there yet. Its doing everything other than building the bundled Mesquite. I've got to work out what CMake command is necessary to add to `bundled/mesquite-2.99/CMakeLists.txt` to prevent this build error from occurring:
```bash
<path_to>/deal.II/dealii/bundled/mesquite-2.99/src/include/MeshInterface.hpp:43:10: fatal error: 'TopologyInfo.hpp' file not found
#include "TopologyInfo.hpp"
```
It looks to me like everything is there, but there's a clearly a little something missing. But that's a job for tomorrow.

**Comment by jppelteret** [comment]
@drwells I finally have a bundled Mesquite building. I'm going to rebuild the library, and then I need to add a small patch to fix something that I know is broken in Mesquite (yes, with the final version that they'll ever ship) and tidy this whole branch up.

@tamiko Can you please tell me what I need to do to prevent the bundled Mesquite from being built if its already detected as a Trilinos module. Trilinos is definitely configured before Mesquite, so can I somehow skip the bundled build if `DEAL_II_TRILINOS_WITH_MESQUITE` is defined?

File: bundled/mesquite-2.99/CMakeLists.txt [hunk 6]
```diff
@@ -26,6 +27,14 @@ STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wformat")
...
 STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wformat-invalid-specifier")
 
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wpointer-arith")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wswitch -Wsynth")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wwrite-strings")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wno-deprecated-declarations")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Qunused-arguments")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wno-unsupported-friend")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wno-parentheses")
+STRIP_FLAG(DEAL_II_CXX_FLAGS "-Wno-unused-local-typedefs")
```
Comment (jppelteret) [inline bundled/mesquite-2.99/CMakeLists.txt:37]: @tamiko I stripped out all of our additional build flags, but somehow they sometimes still find their way into the build command on some of the testers (and on my local machine too). What is the correct way to ensure that the bundled package is built without any warnings being emitted, or warnings being converted to build errors?

**Comment by drwells** [comment]
I think that, at the workshop, we agreed to not bundle this. Is this correct? If so, we should move this to an external repository.

**Comment by bangerth** [comment]
Oh, I remember that we *had* decided to bundle it because it's abandoned upstream. Or do you mean to keep Mesquite in a repository under the dealii organization?

**Comment by drwells** [comment]
Perhaps I misremember what we agreed upon. @jppelteret?

**Comment by jppelteret** [comment]
> we had decided to bundle it because it's abandoned upstream. 

> keep Mesquite in a repository under the dealii organization

We decided to do both these things. The former because we would then make it an unconditional dependency, and the latter so that it can more easily be added to distros.

**Comment by peterrum** [comment]
Maybe we could wrap this PR up after the release?

**Comment by jppelteret** [comment]
We'll go a different route with this; #12444 documents a more modern approach to this problem.

**Pull request closed without merging**